File: tests/client_compatibility/jdbc/src/test/java/com/intersystems/iris/pgwire/SimpleQueryTest.java
package com.intersystems.iris.pgwire;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.BeforeAll;
import static org.junit.jupiter.api.Assertions.*;

import java.sql.*;

/**
 * Test simple query execution via PostgreSQL JDBC driver.
 *
 * Tests P1 Simple Query Protocol:
 * - Query message handling
 * - IRIS SQL execution
 * - Row data encoding
 * - CommandComplete and ReadyForQuery
 */
public class SimpleQueryTest {

    private static String jdbcUrl;
    private static String username;
    private static String password;

    @BeforeAll
    public static void setup() {
        String host = System.getProperty("pgwire.host", "localhost");
        int port = Integer.parseInt(System.getProperty("pgwire.port", "5432"));
        String database = System.getProperty("pgwire.database", "USER");
        username = System.getProperty("pgwire.username", "test_user");
        password = System.getProperty("pgwire.password", "test");

        jdbcUrl = String.format("jdbc:postgresql://%s:%d/%s", host, port, database);
    }

    @Test
    public void testSelectConstant() throws SQLException {
        // GIVEN: Active connection
        try (Connection conn = DriverManager.getConnection(jdbcUrl, username, password);
             Statement stmt = conn.createStatement()) {

            // WHEN: Executing SELECT 1
            ResultSet rs = stmt.executeQuery("SELECT 1");

            // THEN: Should return single row with value 1
            assertTrue(rs.next(), "Result set should have at least one row");
            assertEquals(1, rs.getInt(1), "First column should be 1");
            assertFalse(rs.next(), "Result set should have only one row");
        }
    }

    @Test
    public void testSelectMultipleColumns() throws SQLException {
        // GIVEN: Active connection
        try (Connection conn = DriverManager.getConnection(jdbcUrl, username, password);
             Statement stmt = conn.createStatement()) {

            // WHEN: Executing multi-column SELECT
            ResultSet rs = stmt.executeQuery("SELECT 1 AS num, 'hello' AS text, 3.14 AS float_val");

            // THEN: Should return all columns correctly
            assertTrue(rs.next());
            assertEquals(1, rs.getInt("num"));
            assertEquals("hello", rs.getString("text"));
            assertEquals(3.14, rs.getDouble("float_val"), 0.001);
        }
    }

    @Test
    public void testSelectCurrentTimestamp() throws SQLException {
        // GIVEN: Active connection
        try (Connection conn = DriverManager.getConnection(jdbcUrl, username, password);
             Statement stmt = conn.createStatement()) {

            // WHEN: Executing SELECT CURRENT_TIMESTAMP
            ResultSet rs = stmt.executeQuery("SELECT CURRENT_TIMESTAMP");

            // THEN: Should return a timestamp
            assertTrue(rs.next());
            Timestamp ts = rs.getTimestamp(1);
            assertNotNull(ts, "CURRENT_TIMESTAMP should return a value");
            System.out.println("Current Timestamp: " + ts);
        }
    }

    @Test
    public void testSelectWithNullValue() throws SQLException {
        // GIVEN: Active connection
        try (Connection conn = DriverManager.getConnection(jdbcUrl, username, password);
             Statement stmt = conn.createStatement()) {

            // WHEN: Executing SELECT with NULL
            ResultSet rs = stmt.executeQuery("SELECT NULL AS null_col, 42 AS num_col");

            // THEN: NULL should be handled correctly
            assertTrue(rs.next());
            assertNull(rs.getObject("null_col"), "null_col should be NULL");
            assertTrue(rs.wasNull(), "wasNull() should return true after reading NULL");
            assertEquals(42, rs.getInt("num_col"));
        }
    }

    @Test
    public void testMultipleQueries() throws SQLException {
        // GIVEN: Active connection
        try (Connection conn = DriverManager.getConnection(jdbcUrl, username, password);
             Statement stmt = conn.createStatement()) {

            // WHEN: Executing multiple queries sequentially
            ResultSet rs1 = stmt.executeQuery("SELECT 1");
            assertTrue(rs1.next());
            assertEquals(1, rs1.getInt(1));
            rs1.close();

            ResultSet rs2 = stmt.executeQuery("SELECT 'second query'");
            assertTrue(rs2.next());
            assertEquals("second query", rs2.getString(1));
            rs2.close();

            // THEN: Both queries should succeed
            // (implicit assertion: no exceptions thrown)
        }
    }

    @Test
    public void testResultSetMetadata() throws SQLException {
        // GIVEN: Active connection
        try (Connection conn = DriverManager.getConnection(jdbcUrl, username, password);
             Statement stmt = conn.createStatement()) {

            // WHEN: Executing query and examining metadata
            ResultSet rs = stmt.executeQuery("SELECT 1 AS id, 'test' AS name");
            ResultSetMetaData metadata = rs.getMetaData();

            // THEN: Metadata should describe result columns
            assertEquals(2, metadata.getColumnCount(), "Should have 2 columns");
            assertEquals("id", metadata.getColumnName(1).toLowerCase());
            assertEquals("name", metadata.getColumnName(2).toLowerCase());
            assertEquals(Types.INTEGER, metadata.getColumnType(1));
            assertEquals(Types.VARCHAR, metadata.getColumnType(2));

            System.out.println("Column 1: " + metadata.getColumnName(1) + " (type: " + metadata.getColumnTypeName(1) + ")");
            System.out.println("Column 2: " + metadata.getColumnName(2) + " (type: " + metadata.getColumnTypeName(2) + ")");
        }
    }

    @Test
    public void testEmptyResultSet() throws SQLException {
        // GIVEN: Active connection and a table with no rows
        try (Connection conn = DriverManager.getConnection(jdbcUrl, username, password);
             Statement stmt = conn.createStatement()) {

            // Create temporary table
            stmt.execute("CREATE TABLE IF NOT EXISTS test_empty (id INT)");

            try {
                // Delete all rows to ensure empty
                stmt.execute("DELETE FROM test_empty");

                // WHEN: Querying empty table
                ResultSet rs = stmt.executeQuery("SELECT * FROM test_empty");

                // THEN: Result set should be empty
                assertFalse(rs.next(), "Result set should be empty");
            } finally {
                // Cleanup
                stmt.execute("DROP TABLE IF EXISTS test_empty");
            }
        }
    }
}

File: tests/client_compatibility/dotnet/SimpleQueryTest.cs
using Npgsql;
using Xunit;

namespace PGWireCompatibility;

/// <summary>
/// Test simple query execution via Npgsql driver.
///
/// Tests P1 Simple Query Protocol:
/// - Query message handling
/// - IRIS SQL execution
/// - Row data encoding
/// - CommandComplete and ReadyForQuery
/// </summary>
public class SimpleQueryTest
{
    private readonly string connectionString;

    public SimpleQueryTest()
    {
        var host = Environment.GetEnvironmentVariable("PGWIRE_HOST") ?? "localhost";
        var port = Environment.GetEnvironmentVariable("PGWIRE_PORT") ?? "5432";
        var database = Environment.GetEnvironmentVariable("PGWIRE_DATABASE") ?? "USER";
        var username = Environment.GetEnvironmentVariable("PGWIRE_USERNAME") ?? "test_user";
        var password = Environment.GetEnvironmentVariable("PGWIRE_PASSWORD") ?? "test";

        connectionString = $"Host={host};Port={port};Database={database};Username={username};Password={password}";
    }

    [Fact]
    public async Task TestSelectConstant()
    {
        // GIVEN: Active connection
        await using var conn = new NpgsqlConnection(connectionString);
        await conn.OpenAsync();

        // WHEN: Executing SELECT 1
        await using var cmd = new NpgsqlCommand("SELECT 1", conn);
        var result = await cmd.ExecuteScalarAsync();

        // THEN: Should return 1
        Assert.NotNull(result);
        Assert.Equal(1, Convert.ToInt32(result));
    }

    [Fact]
    public async Task TestSelectMultipleColumns()
    {
        // GIVEN: Active connection
        await using var conn = new NpgsqlConnection(connectionString);
        await conn.OpenAsync();

        // WHEN: Executing multi-column SELECT
        await using var cmd = new NpgsqlCommand("SELECT 1 AS num, 'hello' AS text, 3.14 AS float_val", conn);
        await using var reader = await cmd.ExecuteReaderAsync();

        // THEN: Should return all columns correctly
        Assert.True(await reader.ReadAsync());
        Assert.Equal(1, reader.GetInt32(0));
        Assert.Equal("hello", reader.GetString(1));
        Assert.Equal(3.14, reader.GetDouble(2), precision: 2);
        Assert.False(await reader.ReadAsync());
    }

    [Fact]
    public async Task TestSelectCurrentTimestamp()
    {
        // GIVEN: Active connection
        await using var conn = new NpgsqlConnection(connectionString);
        await conn.OpenAsync();

        // WHEN: Executing SELECT CURRENT_TIMESTAMP
        await using var cmd = new NpgsqlCommand("SELECT CURRENT_TIMESTAMP", conn);
        var result = await cmd.ExecuteScalarAsync();

        // THEN: Should return a timestamp
        Assert.NotNull(result);
        Assert.IsType<DateTime>(result);
        Console.WriteLine($"Current Timestamp: {result}");
    }

    [Fact]
    public async Task TestSelectWithNullValue()
    {
        // GIVEN: Active connection
        await using var conn = new NpgsqlConnection(connectionString);
        await conn.OpenAsync();

        // WHEN: Executing SELECT with NULL
        await using var cmd = new NpgsqlCommand("SELECT NULL AS null_col, 42 AS num_col", conn);
        await using var reader = await cmd.ExecuteReaderAsync();

        // THEN: NULL should be handled correctly
        Assert.True(await reader.ReadAsync());
        Assert.True(reader.IsDBNull(0));
        Assert.Equal(42, reader.GetInt32(1));
    }

    [Fact]
    public async Task TestMultipleQueries()
    {
        // GIVEN: Active connection
        await using var conn = new NpgsqlConnection(connectionString);
        await conn.OpenAsync();

        // WHEN: Executing multiple queries sequentially
        await using var cmd1 = new NpgsqlCommand("SELECT 1", conn);
        var result1 = await cmd1.ExecuteScalarAsync();
        Assert.Equal(1, Convert.ToInt32(result1));

        await using var cmd2 = new NpgsqlCommand("SELECT 'second query'", conn);
        var result2 = await cmd2.ExecuteScalarAsync();
        Assert.Equal("second query", result2);

        // THEN: Both queries should succeed
        // (implicit assertion: no exceptions thrown)
    }

    [Fact]
    public async Task TestExecuteNonQuery()
    {
        // GIVEN: Active connection with test table
        await using var conn = new NpgsqlConnection(connectionString);
        await conn.OpenAsync();

        await using var createCmd = new NpgsqlCommand("CREATE TABLE IF NOT EXISTS test_nonquery (id INT)", conn);
        await createCmd.ExecuteNonQueryAsync();

        try
        {
            // WHEN: Executing DELETE
            await using var deleteCmd = new NpgsqlCommand("DELETE FROM test_nonquery", conn);
            var rowsAffected = await deleteCmd.ExecuteNonQueryAsync();

            // THEN: Should return number of rows affected (may be 0)
            Assert.True(rowsAffected >= 0);
        }
        finally
        {
            // Cleanup
            await using var dropCmd = new NpgsqlCommand("DROP TABLE IF EXISTS test_nonquery", conn);
            await dropCmd.ExecuteNonQueryAsync();
        }
    }

    [Fact]
    public async Task TestDataReaderColumnMetadata()
    {
        // GIVEN: Active connection
        await using var conn = new NpgsqlConnection(connectionString);
        await conn.OpenAsync();

        // WHEN: Executing query and examining metadata
        await using var cmd = new NpgsqlCommand("SELECT 1 AS id, 'test' AS name", conn);
        await using var reader = await cmd.ExecuteReaderAsync();

        // THEN: Metadata should describe result columns
        Assert.Equal(2, reader.FieldCount);
        Assert.Equal("id", reader.GetName(0).ToLower());
        Assert.Equal("name", reader.GetName(1).ToLower());

        Console.WriteLine($"Column 0: {reader.GetName(0)} (type: {reader.GetDataTypeName(0)})");
        Console.WriteLine($"Column 1: {reader.GetName(1)} (type: {reader.GetDataTypeName(1)})");
    }

    [Fact]
    public async Task TestEmptyResultSet()
    {
        // GIVEN: Active connection and a table with no rows
        await using var conn = new NpgsqlConnection(connectionString);
        await conn.OpenAsync();

        await using var createCmd = new NpgsqlCommand("CREATE TABLE IF NOT EXISTS test_empty (id INT)", conn);
        await createCmd.ExecuteNonQueryAsync();

        try
        {
            // Delete all rows to ensure empty
            await using var deleteCmd = new NpgsqlCommand("DELETE FROM test_empty", conn);
            await deleteCmd.ExecuteNonQueryAsync();

            // WHEN: Querying empty table
            await using var selectCmd = new NpgsqlCommand("SELECT * FROM test_empty", conn);
            await using var reader = await selectCmd.ExecuteReaderAsync();

            // THEN: Result set should be empty
            Assert.False(await reader.ReadAsync());
        }
        finally
        {
            // Cleanup
            await using var dropCmd = new NpgsqlCommand("DROP TABLE IF EXISTS test_empty", conn);
            await dropCmd.ExecuteNonQueryAsync();
        }
    }

    [Fact]
    public async Task TestStringWithSpecialCharacters()
    {
        // GIVEN: Active connection
        await using var conn = new NpgsqlConnection(connectionString);
        await conn.OpenAsync();

        // WHEN: Selecting string with quotes and special characters
        var testString = "O'Reilly's \"Book\"";
        await using var cmd = new NpgsqlCommand($"SELECT '{testString.Replace("'", "''")}'", conn);
        var result = await cmd.ExecuteScalarAsync();

        // THEN: String should be returned correctly
        Assert.Equal(testString, result?.ToString());
    }
}
